rasterio/__init__.py

```diff
@@ -261,13 +261,13 @@ def fp_writer(fp):
             # be taken over by the dataset's context manager if it is not
             # None.
             if mode == 'r':
-                s = DatasetReader(fp)
+                s = DatasetReader(fp, driver=driver, **kwargs)
```

Comment:
This is where the driver and open options get to the dataset class constructor.

Reply:
Thanks for expository comments. I had made it just a bit past noticing in this line that the driver wasn't being passed to `DatasetReader` when I saw you open the PR.